### ISSUE TITLE
Removing items by id

### ISSUE THREAD

**Issue by priyadarshy** [lucaong/minisearch#45]
I've got documents that look like this:

```
const tasks = [
{ id: 1, title: "clean the house" }, 
{ id: 2, title: "eat food" }
]
```

If the title of the task with `id = 1` changes, I'd like to update that change in the index. However, in my current application, I don't have access to the entire old version of the document. I just know the `id` and the new values for the title field. 

In order to remove an item in Minisearch, it looks like I need to pass the whole document that I originally added. Is there a way I can remove an item by `id`? If so, I can just remove by `id` and then add the new document.

**Comment by lucaong** [comment]
Hi @priyadarshy ,
at the moment, deletion is only possible by providing the full document. That is because, in order to delete a document, `MiniSearch` has to “de-index” it, basically performing the inverse operation of indexing it. Deletion by ID could be implemented if `MiniSearch` stored the original documents, but that would mean holding on to them in memory. To avoid paying this cost in use cases when saving the original documents is not necessary, this is left as an implementation choice for the user.

What you can do, is store a mapping of IDs to documents yourself. An efficient way to create such a map is:

```javascript
docById = documents.reduce((byId, document) => {
  byId[document.id] = document
  return byId
}, {})

# Now you can get a document from its ID with:
docById[id]
```

I know that it would be convenient to delete by ID, but it would impose higher memory usage for all, even those who don’t need it, and as shown here the user code to map documents by ID is quite small.

I hope this helps :)

**Comment by lucaong** [comment]
I'll close this, as it should provide a solution for your case, but feel free to comment if necessary.

**Comment by priyadarshy** [comment]
Thank you for the detailed help and explanation of why this isn't supported. Your decision making makes total sense to me. I'm happy to implement this within my application in order support removals, I was just avoiding it _incase_ there was a way to remove items by `id`. 

Thank you for building this awesome library. Before I implemented minisearch, I experimented with fuse, lunr and flexsearch. Minisearch had the best documentation and the it's search UX felt like it worked nicely out of the box, like a user would expect in most web apps. I found the rest of libraries to be very academic and hard to get started with. Thank you!

**Comment by lucaong** [comment]
Thanks a lot for the kind words, I am happy you are finding `MiniSearch` useful!

**Comment by lukevp** [comment]
@lucaong I had to go back to FlexSearch because of this - I am streaming documents in from an external system that supports multi-way replication, and I can't reliably detect the update events in a way that I can retrieve and rebuild the indexed document before the change is applied.

I understand your stance on not providing this as default behavior, but I do want to put another vote on this as a feature that MiniSearch could provide that would have enabled my use case.  I.E. if I could pass a config option to allowRemovalByID=true, then it could track those IDs internally as they were set, that would be extremely useful and I would not mind the extra memory usage just for this feature as I was opting in to it.  @clibu and I had a discussion about this on the flexsearch repo.

**Comment by lucaong** [comment]
Hi @lukevp 
thanks for your comment.

In order to support removing a document by ID, `MiniSearch` would have to do exactly the same as the small snippet of code that I posted above. I don't see a case in which `MiniSearch` could do that, but the user code could not (please correct me with an example in case I am missing something). Basically, one would wrap the `add`, `addAll` functions and implement a `removeById` function like this:

```javascript
const miniSearch = new MiniSearch(...)
const docById = {}

const addAll = (documents) => {
  documents.forEach((document) => {
    docById[document.id] = document
  })

  miniSearch.addAll(documents)
}

const add = (document) => {
  docById[document.id] = document
  miniSearch.add(document)
}

const removeById = (id) => {
  const document = docById[id]
  miniSearch.remove(document)
  delete docById[id]
}
```

I am skeptical about adding an option that could be so easily composed by user code on top of `MiniSearch`.

**Comment by mashpie** [comment]
Hey @lucaong, @lukevp - just wanted to jump in to point out and clarify, because I see a potential miss-understanding I stumbled upon recently too.

`miniSearch` __can__ remove single documents by

```js
miniSearch.remove(document)
```
  
`miniSearch` __cannot__ remove single documents by

```js
miniSearch.remove(document.id)
```

and that's it.

---

Most use cases will require some kind of id->doc mapping anyways. Having a list of userProfiles for example and a search limited to names and emails only would lead me to implement a `miniSearch` with those fields only plus another map to reference a whole profile by it's id.

So I end up with a very small index providing very fast search results. Using ids from result to build a list of matching profiles by fetching them from my map. Now that map serves both: keeping details and enable removing from index.

**Comment by lucaong** [comment]
absolutely correct @mashpie, thanks for the clarification.

**Comment by rfox12** [comment]
@lukevp I understand you're new to search, but surely you can understand the memory implications of what you're asking for.  If you **really** want the easy button here's what you can do: **_store the document in the index._**

```
const miniSearch = new MiniSearch({
  fields: ['title', 'author.name', 'tags'], // fields to index for full-text search
  storeFields: ['serializedDoc'], // fields to return with search results
 });

var doc = { id: 1, title: 'Moby Dick', author: { name: 'Herman Melville' }, tags: ['fiction', 'whale'] };
doc.serializedDoc = JSON.stringify(doc);

miniSearch.add(doc);
...
```
And then when you get a document just use JSON.parse() to deserialize your doc, and then remove it from the index.

MiniSearch rocks!

**Comment by lukevp** [comment]
@rfox12  I'm not new to using search engines, just in-browser JS search engines.  As stated above, Elasticsearch and Solr both support removal by ID only, so it doesn't feel like an unreasonable expectation to have that a search engine could support this functionality.

The memory implications of what I'm suggesting are the same regardless of whether MiniSearch supports this or it's built out separately, so the question is just whether this is reasonable functionality for a search engine to support.  I believe it is a common enough use-case that it should be available as an option.  There are also further potential optimizations that could be done within the search engine to support this entity tracking and removal besides using the entire data in memory that could lower the memory usage of this mode (for example, using key compression techniques or hashing), and it could be abstracted from the user if it was part of the public interface of the search engine.

That's an interesting strategy for tracking the documents within the index, thanks for the info.

**Comment by mashpie** [comment]
often writing code is faster than arguing, so...

This should be a valid "drop-in" extending miniSearch with `miniSearch.removeById()` - rest of api is inherited from minisearch.

```js
import MiniSearch from 'minisearch'

export default class MySearch extends MiniSearch {
  constructor(options) {
    super(options)
    this.docById = new Map()
  }

  add(document) {
    this.docById.set(document.id, document)
    super.add(document)
  }

  addAll(documents) {
    documents.forEach((document) => this.add(document))
  }

  removeById(id) {
    super.remove(this.docById.get(id))
    this.docById.delete(id)
  }
}
```

if you need to support IE10 or lower you should add a `Map()` polyfill or change to  `this.docById = {}`

---

btw. this is close to how I use minisearch in some projects to return the full document on search:

```js
import MiniSearch from 'minisearch'

export default class MySearch extends MiniSearch {
  constructor(options) {
    super(options)
    this.docById = new Map()
  }

  add(document) {
    this.docById.set(document.id, document)
    super.add(document)
  }

  addAll(documents) {
    documents.forEach((document) => this.add(document))
  }

  search(query) {
    return super.search(query).map((meta) => this.docById.get(meta.id))
  }
}
```

**Comment by lukevp** [comment]
@mashpie  I appreciate the continued help you are providing.  However, my problem is not an inability to implement this within my own code, it is a desire to provide feedback on a feature I feel is missing from this library.  I do feel that the core statement I am trying to make is getting lost so I will restate it as clearly as I can one last time.

- Other search engines support removal by ID.  
- MiniSearch is a search engine, therefore it could support this feature, as others do.
- This is an important use case for me and others as well.  
- I feel this feature could be enabled optionally by end users of the library, but not impact users who do not need this behavior.
- It appears that you all consider this outside the scope of MiniSearch.  That is fine and is your decision to make.  

I'm not here to argue, but to provide a viewpoint that other potential users other than myself also share, that feeds into our decision to use or not use this library.  I feel that I have provided enough reasoning for my stance and it has become clear that this feature is not under consideration, so I'll use a different library.  No big deal and thank you all for being so responsive.  It seems like a great community, I just have a different viewpoint on this issue.

**Comment by clibu** [comment]
I'll either add to the clarity or make it worse.😀

The issue with minisearch has nothing to do with document id's as such. Where minisearch differs from most if not all other full text search engines is that in order to remove a document from the index, the document that was added is required. In other words, it is the content of that original document that is parsed and removed from the index. 

This carries over to updating a document, in that you can't just update one. You first need to remove the _original document_ and then add the _updated document_.

This differs from other fts libraries I am aware of. Further many fts libraries actually store the documents being indexed, which would mean they could handle this remove/update process if that's how they worked. FlexSearch does store the documents, but not for this purpose.

In my use case and I assume many others, I don't want the overhead of the fts library storing the documents _(either in memory or on disk)_, especially when we are talking about Browser resources, which are constrained. In my app I already have the documents stored in IndexedDB, so why would I want to double up. There is a FlexSearch [issue](https://github.com/nextapps-de/flexsearch/issues/155) on just this.

In order to meet minisearch's requirements I did have to rework my code to get a copy of the current document from the database, before updating it and pass the _before_ and _after_  documents to a function to update minisearch. I was able to do this with a bit of mucking around, as document updates can come from the local user or over the wire from other users. I can however see this could be problematic for some.

**Comment by mashpie** [comment]
All good and fine. Other engines I tried (lunr, fuse, elasticlunr) either took too long for indexing or consumed too much memory (yes, by keeping copies of source docs). Removal and update operations always means to at least partly rebuild the existing index. Even flexsearch performs .remove() and .add() on .update() as far as I can see in code.

We all choose whatever best fits our needs and those vary. To me the lack of an „internal“ copy is a big benefit as I don‘t waste memory. And in 90% of my use cases I even don‘t need to remove any docs but want fastest indexing as soon a new collection pushes in from server.

Another use case tracks changes of single docs stored by IndexDB. Having the source in there I don‘t need another copy in memory and for deletion I just .remove(docFromIndexDB).

So, yes I really appreciate that removeById is missing in minisearch and that minisearch trys to focus on that current niche (just terms, a bit of fuzzy, custom scoring but no wildcard, grouping, sorting, aggregation) - and yes I personally prefer to add a custom feature than using libs that carry tons of unwanted extra features... there is a sweet spot in balance.

I use a hammer and a screwdriver at home, but on the road I carry a multitool. Which one is better? Both in there use case :) - Ok, sorry I didn‘t want to argue, just couldn‘t resist that discussion. 

Stay safe and have nice weekend!

**Comment by lucaong** [comment]
@mashpie @clibu @lukevp @rfox12 @priyadarshy I revive this old thread to announce that the newly released `v6.0.0-beta.1` (soon to be `v6.0.0`) includes a new method `discard` (and related methods `discardAll` and `replace`) which removes a document by ID.

The reasons given in this thread for why `remove` takes a full document are still valid, but `discard` takes a different approach, made possible by the internal design of `MiniSearch`. Instead of immediately removing a document from the index (which would only be feasible given the original document), `discard` marks the current document version as discarded, so it is immediately ignored by searches, but does not mutate the index. Instead, that is left to a vacuuming process, ran by default after a certain number of documents are discarded. Vacuuming traverses the index and removes references to discarded documents, allowing memory to be released. See the documentation here: https://lucaong.github.io/minisearch/classes/_minisearch_.minisearch.html#discard

This provides the feature requested by this issue, while avoiding the drawbacks of a naive implementation: the original documents are not stored, and use cases that do not need to remove documents incur no overhead.

**Comment by clibu** [comment]
@lucaong This is excellent news. Thanks so much for adding this much needed capability. I look forward to testing the new release.👏👏😁
